File: Data Structures & Algorithms/max-water-container/submission-3.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        left, right = 0, len(heights) - 1
        max_area = 0

        while left < right:
            current_max = (right - left) * min(heights[left], heights[right])

            if heights[left] < heights[right]:
                left += 1
                if current_max > max_area:
                    max_area = current_max

            else:
                right -= 1
                if current_max > max_area:
                    max_area = current_max
        
        return max_area        
                                 
        

File: Data Structures & Algorithms/eating-bananas/submission-3.java
class Solution {

    public int calcMax(int[] piles) {
        int currentMax = 0;
        for(int pile: piles) {
            currentMax = Integer.max(currentMax, pile);
        }
        return currentMax;
    }

    public int minEatingSpeed(int[] piles, int h) {

        int left = 1, right = calcMax(piles);

        while(left <= right) {
            int middle = left + (right - left) / 2, totalHours = 0;
            for(int pile : piles) {
                totalHours += (pile + middle - 1) / middle;
            }
            if(totalHours <= h) {
                right = middle - 1;
            }
            else {
                left = middle + 1;
            }
        }
        return left;
    }
}
